packages/phase/README.md

```diff
@@ -0,0 +1,134 @@
+# @loopback/phase
+
+This module defines `Phase` and `PhaseList` to organize a list of handlers by
+groups. It's rewritten in TypeScript based on
+https://github.com/strongloop/loopback-phase.
+
+## Installation
+
+```sh
+npm install --save @loopback/phase
+```
+
+## Basic Use
+
+### Handler
+
+A handler is a function that takes a `Context` object, an optional
+`HandlerChain` object and returns a `Promise`.
+
+When a handler is invoked as part of a chain, it can control how to proceed as
+follows:
+
+- Continue to invoke downstream handlers after it exits
+
+```ts
+async ctx => {
+  console.log(ctx.req.url);
+};
+```
+
+- Invoke downstream handlers within the method in cascading style
+
+```ts
+async (ctx, chain) => {
+  const start = process.hrtime();
+  await chain!.next();
+  const duration = process.hrtime(start);
+  console.log('Duration: ', duration);
+};
+```
+
+- Terminate the handler chain and return
+
+```ts
+async (ctx, chain) => {
+  if (ctx.req.url === '/status') {
+    ctx.response = {status: 'ok'};
+    chain!.return();
+  }
+};
+```
+
+- Abort the handler chain by throwing an error
+
+```ts
+async (ctx, chain) => {
+  throw new Error('invalid request');
+  // or
+  // chain!.throw(new Error('invalid request'));
+};
+```
+
+### Phase
+
+A `Phase` is a bucket for organizing handlers. Each phase has an `id` and three
+sub-phases:
+
+- before (contains handlers to be invoked before the phase)
+- use (contains handlers to be invoked during the phase)
+- after (contains handlers to be invoked after the phase)
+
+The handlers within the same subphase will be executed in serial or parallel
+depending on the `options.parallel` flag, which is default to `false`.
+
+The three sub-phases within the same phase will be executed in the order of
+`before`, `use`, and `after`.
+
+There is also a `failFast` option to control how to handle errors. If `failFast`
+is set to `false`, errors will be caught and set on the `ctx.error` without
+aborting the handler chain.
```

Comment:
I find this rather dangerous. I expect that most people writing phase-handlers are not going to be aware of the `failFast:false` mode. As a result, `failFast:false` mode can execute handlers that are not prepared to handle the situation left after one of the previous handlers failed.

I think a better solution would be to enable `failFast:false` mode on a per-handler basis - i.e. allow each handler to signal whether it wants to be executed even if there was an error up in the chain.

What is the use case you are envisioning for `failFast:false` mode?

Can we perhaps leave this feature out of this initial pull request, to get it landed faster?

Reply:
I introduce `failFast:false` mode mostly for error phase so that we can have a list of error handlers to mediate errors.